### DISCUSSION TITLE
Best way to return a "success" text for a @LuaFunction?

### DISCUSSION THREAD

**Discussion by Lordmau5** [cc-tweaked/CC-Tweaked#1311] [Q&A] [answered]
Hey there!

I'm currently working on implementing CC support in my mod [FFS](https://github.com/Lordmau5/FFS) again and was curious about something.

I already figured out that I can throw a `LuaException` when something isn't as it should, e.g. my tank being invalid.

I wonder if there's something similar for a method to return that it succeeded?

For example, I have a method to lock the fluid in the tank and when it locks it I would like to return a string along the lines of "Locked fluid to [FLUID_NAME].".

If this is not intended and I should rather just return a `boolean` on whether it succeeded or not I'd be fine with that, but I just wasn't sure what the commonly accepted way for this is :)

**Comment by SirEdvin** [comment, marked as answer]
I am not sure, that there is a commonly accepted way for this, but I can suggest to use something like [Result Monad](https://adambennett.dev/2020/05/the-result-monad/), which is effectefly just ~~returning table `{isOperationSuccessful, problemWithOperation}`~~ and return exception only in cases of invalid output of some incorrect usage of function

Update: **Not a table**, just one or two values for easly handle like in golang:

```lua
local result, err = true, "haha"

if err ~= nil then
print(err)
end
```

**Reply by Lordmau5** [reply]
How would I return a table in that case? Do I set my method return type to `LuaTable` and then build a map with the results?

Because that'd be quite nice to have for a different method I have to get the tank info haha

**Reply by SquidDev** [reply]
You can return a `Map<?, ?>` to represent a Lua table. See the docs from `MethodResult`:

https://github.com/cc-tweaked/CC-Tweaked/blob/d066d175bf0de89748d248c5301bd95b88f2d85f/projects/core-api/src/main/java/dan200/computercraft/api/lua/MethodResult.java#L51-L58

and `LuaFunction`:

https://github.com/cc-tweaked/CC-Tweaked/blob/d066d175bf0de89748d248c5301bd95b88f2d85f/projects/core-api/src/main/java/dan200/computercraft/api/lua/LuaFunction.java#L18-L36

Though typically here you wouldn't return a table and would return multiple values (so `new Object[] { success, message }`).

**Reply by Lordmau5** [reply]
Sweet! Thank you guys!

**Reply by SirEdvin** [reply]
I usually just return `MethodResult.of(isOperationSuccessful, problemWithOperation)` like [here](https://github.com/SirEdvin/UnlimitedPeripheralWorks/blob/1.18/src/main/kotlin/site/siredvin/peripheralworks/integrations/ae2/MENetworkBlockPlugin.kt#L399).

Oh, I see know, that I make a huge mistake in my first comment. This is not a table, it is just two values (or single, eventually)

**Reply by SirEdvin** [reply]
Sorry for this mistake :(
